flt.lttng-utils.debug-info: adapt debug-info component to API changes

frdeso opened the pull request:
General idea
------------
Given that:
	* debug-info filter component needs to add new fields to the events
		common context,
	* Metadata and data objects are immutable after being added to the trace
		class or trace respectively.

The debug-info component needs output a new augmented trace containing
the new fields along side all the received input data. It thus needs to
act both as a sink AND a source of graph messages in the sense that it
consumes and creates graph messages.

For every message received, the corresponding output objects(e.g.
stream) are created in an output trace. Those objects are stored in
hashmaps corresponding to their types for later use. These hashmaps map
input object to their corresponding output objects.

For example, when handling a packet beginning message, we need to create
a new output packet. To do that, the corresponding output stream is
needed, amongst other things. This output stream was create earlier when
a stream beginning message was received and can be recovered from the
stream hashmap.

For each type of message received from the upstream iterator, we must
create a copy of that message and its corresponding trace IR objects,
and send it downstream. This is true for all but one message type,
inactivity messages can be forwarded directly as they are independent of
the data and metadata IR objects.

Implementation
--------------
This commit adds the concept of mapping between input and output objects
and implements copy functions to create (nearly) identical output objects
based on the input ones.

The entry point of the mapping of objects is the struct trace_id_maps
structure.

The metadata_map and data_map hashtables have input trace class and input trace
objects, respectively, as key and are pointing to struct trace_ir_metadata_maps and
struct trace_ir_data_maps. Each of the IR object maps contained in these
structs have the input object of their specific types as key.

All input and mapped output objects are identical except for the stream
classes. Event-common-context of the stream classes needs to be modify to
include the debug-info fields if and only if the right fields (instruction
pointer (ip) and vpid) are present in the input stream class.
	Here are the objects kept in the mapping:
	- Trace class
	- Stream class
	- Event class
	- Clock class
	- Trace
	- Stream
	- Packet

The memory layout of each processes is kept up to date using the
lttng-ust-statedump and lib-load events. This information is then used
to resolve source locations based on ip and vpid values found in the
event-common-context of events.

Future Improvements
---------------------
Decoupling the trace-ir-mapping library from the debug-info component so
it can be used in other filter components.

Other fixes
-----------
- Don't resolve debug-info on build id mismatch
- When parsing the build id of a statedump event, confirm that it matches the
	one present on disk for that file.
- Prevent NULL pointer access of arg_debug_info_field_name
- Typo in the lttng_ust_statedump:debug_link event name.
- Typo in the crc field name of the lttng_ust_statedump:debug_link event.

frdeso committed:
flt.lttng-utils.debug-info: adapt debug-info component to API changes

General idea
------------
Given that:
	* debug-info filter component needs to add new fields to the
	  events common context,
	* Metadata and data objects are immutable after being added to
	  the trace class or trace respectively.

The debug-info component needs output a new augmented trace containing
the new fields along side all the received input data. It thus needs to
act both as a sink AND a source of graph messages in the sense that it
consumes and creates graph messages.

For every message received, the corresponding output objects(e.g.
stream) are created in an output trace. Those objects are stored in
hashmaps corresponding to their types for later use. These hashmaps map
input object to their corresponding output objects.

For example, when handling a packet beginning message, we need to create
a new output packet. To do that, the corresponding output stream is
needed, amongst other things. This output stream was created earlier
when a stream beginning message was received and can be recovered from
the stream hashmap.

For each type of message received from the upstream iterator, we must
create a copy of that message and its corresponding trace IR objects,
and send it downstream. This is true for all but one message type,
inactivity messages can be forwarded directly as they are independent of
the data and metadata IR objects.

Implementation
--------------
This commit adds the concept of mapping between input and output objects
and implements copy functions to create (nearly) identical output
objects based on the input ones.

The entry point of the mapping of objects is the struct trace_id_maps
structure.

The metadata_map and data_map hashtables have input trace class and
input trace objects, respectively, as key and are pointing to struct
trace_ir_metadata_maps and struct trace_ir_data_maps. Each of the IR
object maps contained in these structs have the input object of their
specific types as key.

All input and mapped output objects are identical except for the stream
classes. Event-common-context of the stream classes needs to be modified
to include the debug-info fields if and only if the right fields
(instruction pointer (ip) and vpid) are present in the input stream
class.
Here are the objects kept in the mapping:
  - Trace class
  - Stream class
  - Event class
  - Clock class
  - Trace
  - Stream
  - Packet

The memory layout of each processes is kept up to date using the
lttng-ust-statedump and lib-load events. This information is then used
to resolve source locations based on ip and vpid values found in the
event-common-context of events.

Future Improvements
---------------------
Decoupling the trace-ir-mapping library from the debug-info component so
it can be used in other filter components.

Other fixes
-----------
- Don't resolve debug-info on build id mismatch
- When parsing the build id of a statedump event, confirm that it
  matches the one present on disk for that file.
- Prevent NULL pointer access of arg_debug_info_field_name
- Typo in the lttng_ust_statedump:debug_link event name.
- Typo in the crc field name of the lttng_ust_statedump:debug_link
  event.

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

eepp commented on include/babeltrace/common-internal.h:
```diff
@@ -477,4 +477,30 @@ const char *bt_self_message_iterator_status_string(
+		ret = BT_SELF_MESSAGE_ITERATOR_STATUS_NOMEM;
+		break;
+	default:
+		BT_ASSERT(0);
```
We usually use `abort()` for this purpose.

eepp commented on include/babeltrace/common-internal.h:
```diff
@@ -477,4 +477,30 @@ const char *bt_self_message_iterator_status_string(
+		bt_message_iterator_status status)
+{
+	bt_self_message_iterator_status ret;
+	switch (status) {
```
You can just cast with `(int)` directly here. Status codes are purposefully the same:

```
typedef enum bt_self_message_iterator_status {
	BT_SELF_MESSAGE_ITERATOR_STATUS_OK = BT_MESSAGE_ITERATOR_STATUS_OK,
	BT_SELF_MESSAGE_ITERATOR_STATUS_END = BT_MESSAGE_ITERATOR_STATUS_END,
	BT_SELF_MESSAGE_ITERATOR_STATUS_AGAIN = BT_MESSAGE_ITERATOR_STATUS_AGAIN,
	BT_SELF_MESSAGE_ITERATOR_STATUS_ERROR = BT_MESSAGE_ITERATOR_STATUS_ERROR,
	BT_SELF_MESSAGE_ITERATOR_STATUS_NOMEM = BT_MESSAGE_ITERATOR_STATUS_NOMEM,
} bt_self_message_iterator_status;
```

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -5,6 +5,7 @@
  * Copyright (c) 2015 Philippe Proulx <pproulx@efficios.com>
  * Copyright (c) 2015 Antoine Busque <abusque@efficios.com>
  * Copyright (c) 2016 Jérémie Galarneau <jeremie.galarneau@efficios.com>
+ * Copyright (c) 2019 Francis Deslauriers <francis.deslauriers@efficios.com>
```
Welcome to this file :smiley:.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -98,9 +153,96 @@ void debug_info_source_destroy(struct debug_info_source *debug_info_src)
+		/* Get a reference on the input trace. */
+		bt_trace_class_get_ref(in_trace_class);
+		g_hash_table_insert(debug_it->metadata_map,
+				(gpointer) in_trace_class, metadata_mapping);
```
Is this cast needed considering that `gpointer` is `void *`?

Check other locations too if not.

frdeso replied:
I get the following warning when I remove the cast:
```
passing 'const bt_trace_class *' (aka 'const struct bt_trace_class *') to parameter of type 'gpointer' (aka 'void *') discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers```

eepp replied:
Oh yes so it's required to cast the `const` away then, I missed that.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -98,9 +153,96 @@ void debug_info_source_destroy(struct debug_info_source *debug_info_src)
+		const bt_event_class *in_event_class)
+{
+	const bt_trace_class *in_trace_class;
+	in_trace_class = bt_stream_class_borrow_trace_class_const(
```
Collapse declaration and assignment here. Do as such for other similar patterns.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	dbg_info_src = debug_info_query(debug_info, vpid, ip);
+
+	/* Start with an empty string. */
+	curr_field_str = g_string_new(NULL);
```
Could we find a way to avoid memory allocation on this fast path?

frdeso replied:
In all cases, we need a string, empty or not. If it needs to be empty, because debug_info_query() returned NULL, we could use a static char array ("\0") to avoid the allocation. Also the loop right after that could be rewrote into a series of call to ` bt_field_class_structure_borrow_member_by_name_const()`. I will look into all this.

eepp replied:
I don't understand how `bt_field_class_structure_borrow_member_by_name_const()` is related here.

But getting rid of the `GString` and using `bt_field_string_append()` in the `fill_debug_info_*()` functions would solve this. Just don't forget to call `bt_field_string_clear()` initially in `fill_debug_info_field()`.

frdeso replied:
My `bt_field_class_structure_borrow_member_by_name_const()` comment was unrelated to your initial comment, it was just something I realized when looking at your comment. I should not have mentioned it.

frdeso replied:
I made it so that there are no allocation in that loop on the fast path. I will make sure you have a look during the next review.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	out_debug_info_field = bt_field_structure_borrow_member_field_by_name(
+				out_common_ctx_field, debug_info_field_name);
+
+	if (out_debug_info_field) {
```
Is there any way that this could cause some name clash? If the debug info field exists in the input event's common context for example?

frdeso replied:
What should we do the debug-info field is already present? should we append to it? create a debug-info-2 field ? 
here is the metadata part of the previous implementation: https://github.com/eepp/babeltrace/blob/optimalize/plugins/lttng-utils/copy.c#L829

Here is my suggestion:

- If there is a `debug-info` struct field class and any of its regular members (`bin`, `func`, `src`) are present, debug-info should act as a passthrough (don't add field classes to the common context) so not to override the work of a upstream component.

- If there is a `debug-info` struct field class and none of its regular members (`bin`, `func` or `src`) are present , debug-info should add the new field classes (`bin`, `func`, `src`) to the `debug-info` struct.

eepp replied:
I would be more strict than that: if there's a `debug-info` member (actually that's just the default name of the `debug_info_field_name` string), whatever its field class, disable the filter for this whole stream class (copy, but don't do any debug information lookup or modification to the copied stream).

The user always has the choice to name this field another way should there be a name clash.

What do you think?

frdeso replied:
Make sense.
So act as a passthrough if `debug-info` field class already present in the event common context.

eepp replied:
Yes, a passthrough, but still copying because there could be another stream class in the same trace class (now or in the future) with a compatible common event context field class.

frdeso replied:
yes. that's what i meant to say.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	metadata_mapping = get_metadata_mapping_from_input_event_class(
+			debug_it, in_event_class);
+
+	out_event_class = data_mapping_borrow_mapped_event_class( metadata_mapping,
```
:artificial_satellite: :moon: :stars:

frdeso replied:
✔️

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+			in_message, &cs);
+
+	/* Create an event output message. */
+	if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
That's not the correct condition. You need to check if the stream class has a default clock class. Then, and only then, you can call `bt_message_event_borrow_default_clock_snapshot_const()`. You can fail if `bt_message_event_borrow_default_clock_snapshot_const()` returns `BT_CLOCK_SNAPSHOT_STATE_UNKNOWN` for the moment; `bt_message_event_create_with_default_clock_snapshot()` needs to be called when there's a default clock class, and otherwise `bt_message_event_create()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+				in_message, &cs);
+
+	/* Create an packet beginning output message. */
+	if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+				in_message, &cs);
+
+	/* Create an packet end output message. */
+	if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	out_message = bt_message_stream_activity_beginning_create(
+			debug_it->input_iterator, out_stream);
+
+	if (cs_state == BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_KNOWN ) {
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	out_message = bt_message_stream_activity_end_create(
+			debug_it->input_iterator, out_stream);
+
+	if (cs_state == BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_KNOWN ) {
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	 * If both clock snapshots are knowns, create the message using their
+	 * values.
+	 */
+	if (begin_cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN &&
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	 * If both clock snapshots are knowns, create the message using their
+	 * values.
+	 */
+	if (begin_cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN &&
```
See comment in `handle_event_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	case BT_MESSAGE_TYPE_PACKET_END:
+		out_message = handle_packet_end_message(debug_it, in_message);
+		break;
+	case BT_MESSAGE_TYPE_EVENT:
```
I would put this one at the top, just as a hint that it's the one you get the most. Now I don't know if this makes it really more efficient.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	BT_SELF_COMPONENT_PORT_INPUT_MESSAGE_ITERATOR_MOVE_REF(
+		debug_info_iter->msg_iter, upstream_iterator);
+
+	debug_info_iter->debug_info_component = (struct debug_info_component *)
```
No cast needed.

eepp commented on plugins/lttng-utils/trace-ir-data-copy.c:
```diff
@@ -0,0 +1,273 @@
+	bt_trace_status status;
+	const char *trace_name;
+
+	BT_LOGD("Copying content of trace: in-t-addr=%p,out-t-addr=%p",
```
:moon: (`p,o`)

Elsewhere too.

eepp commented on plugins/lttng-utils/trace-ir-data-copy.c:
```diff
@@ -0,0 +1,273 @@
+	status = bt_stream_set_name(out_stream, name);
+	if (status != BT_STREAM_STATUS_OK) {
+		BT_LOGE("Cannot set stream's name: stream-addr=%p, "
+			"name=%s", out_stream, name);
```
Then go to end?

eepp commented on plugins/lttng-utils/trace-ir-data-copy.c:
```diff
@@ -0,0 +1,273 @@
+	status = bt_trace_set_name(out_trace, trace_name);
+	if (status != BT_TRACE_STATUS_OK) {
+		BT_LOGE("Cannot set trace's name: trace-addr=%p, name=\"%s\"",
+				out_trace, trace_name);
```
Then go to end?

eepp commented on plugins/lttng-utils/trace-ir-metadata-field-class-copy.c:
```diff
@@ -0,0 +1,677 @@
+#include "trace-ir-metadata-field-class-copy.h"
+
+/*
+ * This fonction walks througth the nested structures field class to resovle a
```
through, resolve

:nose:  :bread: ?

eepp commented on plugins/lttng-utils/trace-ir-metadata-field-class-copy.c:
```diff
@@ -0,0 +1,677 @@
+
+	fp_index_count = bt_field_path_get_index_count(fp);
+	curr_fc = fc;
+	for (i = 0; i < fp_index_count; i++) {
```
Yes that's the spirit.

eepp commented on plugins/lttng-utils/trace-ir-metadata-field-class-copy.c:
```diff
@@ -0,0 +1,677 @@
+}
+
+static inline
+void field_class_signed_enumeration_copy(
```
Can't you have a single version instead, and just check the type (unsigned or signed) at a few places to call the appropriate function? This is a slow path anyway.

eepp commented on plugins/lttng-utils/trace-ir-metadata-field-class-copy.c:
```diff
@@ -0,0 +1,677 @@
+					ranges, range_index, &lower, &upper);
+
+			BT_LOGD("Copying range in enumeration field class: "
+					"label=%s, lower=%lu, upper=%lu",
```
Use `PRId64` and such to remain moulti-platform.

eepp commented on plugins/lttng-utils/trace-ir-metadata-field-class-copy.c:
```diff
@@ -0,0 +1,677 @@
+		bt_field_class *out_elem_field_class;
+
+		in_elem_field_class =
+			bt_field_class_array_borrow_element_field_class_const(
```
There's something in common to be done for static and dynamic array field classes. Everything that's calling `bt_field_class_array_*()` should be common code.

eepp commented on plugins/lttng-utils/trace-ir-mapping.h:
```diff
@@ -0,0 +1,120 @@
+	struct field_class_resolving_context *fc_resolving_ctx;
+
+	/* input_trace -> data_mapping. */
+	GHashTable *data_map;
```
Why does this exist? Could you just put the three data mapping hash tables here instead?

Also, I don't get why data mappings are inside a metadata mapping structure. To me, all those hash tables should be side to side within the iterator's privata data structure. I might be missing something.

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+
+	/* Use the same stream class ids as in the origin trace class. */
+	bt_trace_class_set_assigns_automatic_stream_class_id(out_trace_class,
+			FALSE);
```
`BT_FALSE`

Check other locations too.

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+					out_trace_class, value_name,
+					bt_value_string_get(value));
+		} else {
+			BT_LOGE_STR("Wrong bt_value type.");
```
Should never happen. You can just `abort()`.

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+
+	clock_class_description = bt_clock_class_get_description(in_clock_class);
+
+	status = bt_clock_class_set_description(out_clock_class,
```
`clock_class_description` can be `NULL` (optional property).

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+
+	clock_class_name = bt_clock_class_get_name(in_clock_class);
+
+	status = bt_clock_class_set_name(out_clock_class, clock_class_name);
```
`clock_class_name` can be `NULL` (optional property).

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+	bt_clock_class_set_is_absolute(out_clock_class,
+			bt_clock_class_is_absolute(in_clock_class));
+	bt_clock_class_set_uuid(out_clock_class,
+			bt_clock_class_get_uuid(in_clock_class));
```
`bt_clock_class_get_uuid()` can return `NULL` (optional property).

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+
+	if (bt_field_class_integer_get_field_value_range(vpid_fc) != 32) {
+		match = false;
+		goto end;
```
How about `goto nomatch` for all those cases?

Or `goto end`, but make `match` `false` by default and set it to `true` when you don't go to `end`.

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+	return out_clock_class;
+}
+
+void copy_stream_class_content(struct metadata_mapping *metadata_mapping,
```
A stream class has an optional name too to be copied (see `bt_stream_class_get_name()`).

eepp commented on plugins/lttng-utils/trace-ir-metadata-copy.c:
```diff
@@ -0,0 +1,576 @@
+	}
+
+	bt_trace_class_set_uuid(out_trace_class,
+			bt_trace_class_get_uuid(in_trace_class));
```
`bt_trace_class_get_uuid()` can return `NULL` (optional property).

eepp commented on plugins/lttng-utils/trace-ir-mapping.c:
```diff
@@ -0,0 +1,600 @@
+	if (!metadata_mapping->fc_resolving_ctx) {
+		BT_LOGE_STR("Error allocating field_class_resolving_context");
+		g_free(metadata_mapping->debug_info_field_class_name);
+		g_free(metadata_mapping);
```
I suggest you have an `error` label instead of repeating those `g_free()` calls. The error path should simply call `metadata_mapping_destroy()` if `metadata_mapping` is not `NULL`, but then `metadata_mapping_destroy()` must check that each member is not `NULL` before freeing/destroying it.

eepp commented on plugins/lttng-utils/trace-ir-mapping.c:
```diff
@@ -0,0 +1,600 @@
+		bt_self_component *self_comp, const char *debug_info_field_name)
+{
+	struct metadata_mapping *metadata_mapping
+		= g_new0(struct metadata_mapping, 1);
```
`=` above

eepp commented on plugins/lttng-utils/trace-ir-mapping.c:
```diff
@@ -0,0 +1,600 @@
+	/* There should never be a mapped packet. */
+	BT_ASSERT(!borrow_mapped_packet(metadata_mapping, in_packet));
+
+	in_stream = bt_packet_borrow_stream_const(in_packet);
```
`bt_packet_borrow_stream_const()` is repeated: maybe you should just set `in_stream` and use it at both places?

eepp commented on plugins/lttng-utils/Makefile.am:
```diff
@@ -11,6 +10,14 @@ libdebug_info_la_SOURCES = \
+	trace-ir-metadata-copy.c \
+	trace-ir-metadata-copy.h \
+	trace-ir-metadata-field-class-copy.c \
+	trace-ir-metadata-field-class-copy.h \
```
Good idea to split to multiple files here.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -248,6 +392,106 @@ struct proc_debug_info_sources *proc_debug_info_sources_ht_get_entry(
 }
 
+static inline
+const bt_field *event_get_payload_field(const bt_event *event,
```
`get` :arrow_forward: `borrow`? Other locations too.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -248,6 +392,106 @@ struct proc_debug_info_sources *proc_debug_info_sources_ht_get_entry(
+
+		(*build_id)[i] = bt_field_unsigned_integer_get_value(curr_field);
+	}
+error:
```
Should be named `end` because you reach it without any error too.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -517,46 +727,32 @@ void handle_bin_info_event(FILE *err, struct debug_info *debug_info,
-		is_pic = (tmp == 1);
+		event_get_payload_unsigned_integer_field_value(event,
+				IS_PIC_FIELD_NAME, &is_pic_field_value);
+		is_pic = (is_pic_field_value == 1);
```
Useless parentheses.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		 * empty.
+		 */
+		if (dbg_info_src) {
+			if (!strcmp(curr_field_name, "bin")) {
```
Use `strcmp(...) == 0` just to remain consistent. `!strcmp()` looks negative, even if it's equivalent.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+				fill_debug_info_src_field(dbg_info_src,
+						debug_info->comp->arg_full_path,
+						curr_field_str);
+			}
```
Else abort?

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+
+		bool_val = bt_value_bool_get(value);
+
+		debug_info_component->arg_full_path = bool_val;
```
How about `= bt_value_bool_get(value)` here?

eepp reviewed (changes requested):
This is my first pass.

Are you planning to handle the _is static_ property and listeners?

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +937,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	upstream_iterator_ret_status =
+		bt_self_component_port_input_message_iterator_next(
+				upstream_iterator, &local_msgs, &local_count);
+
```
Just noticed this, but you must check `upstream_iterator_ret_status` before reading `local_count`: there's no guarantee that `local_count` is still 0 if `upstream_iterator_ret_status` is not `BT_MESSAGE_ITERATOR_STATUS_OK`.

frdeso committed:
lib: replace trace is_static with destruction listeners

General idea
------------
Replace bt_trace is_static infrastructure with the concept of
destruction listener and extend it for the bt_trace_class object.

This allows for downstream components to ask to be notify of the
destruction of those objects. This notification is useful for component
that keep data related trace IR objects that should be deallocated when
the trace and/or trace class are destroyed.

This feature is mostly useful to handle sources which could create and
release lots of trace class and trace objects. In this case, when the
source message iterator is done with the stream objects (emitted all the
relevant stream end messages), it puts the stream, trace, and trace
class references. Then downstream users are notified that specific
external objects are about to be destroyed, so they release/close any
associated resource.

The flt.lttng-utils.debug-info component will make use of this
capability to cleanup the mapping between input and output objects it
needs to manage to do its function.

Implementation
--------------
Both trace class and trace offer the same interface and are implemented
in the exact same way.

Destruction listener callbacks and their associated private pointer are
stored in an array and are all called at the execution of the destroy
functions. Destruction listeners are registered using the following
functions:

  bt_trace_status bt_trace_add_destruction_listener(
          const bt_trace *trace,
          bt_trace_destruction_listener_func listener,
          void *data,
          uint64_t *listener_id);

  bt_trace_class_status bt_trace_class_add_destruction_listener(
          const bt_trace_class *trace,
          bt_trace_class_destruction_listener_func listener,
          void *data,
          uint64_t *listener_id);

It's possible to remove a destruction listener using the destruction
listener id received when register it initially used the following
functions:

  bt_trace_status bt_trace_remove_destruction_listener(
          const bt_trace *trace,
          uint64_t listener_id);

  bt_trace_class_status bt_trace_class_remove_destruction_listener(
          const bt_trace_class *trace_class,
          uint64_t listener_id);

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

eepp commented on lib/trace-ir/trace-class.c:
```diff
@@ -63,6 +68,26 @@ void destroy_trace_class(struct bt_object *obj)
 	BT_LIB_LOGD("Destroying trace class object: %!+T", tc);
+	/*
+	 * Call destroy listener functions so that everything else
+	 * still exists in the trace.
```
in the _trace class_

eepp commented on lib/trace-ir/trace-class.c:
```diff
@@ -63,6 +68,26 @@ void destroy_trace_class(struct bt_object *obj)
+	 * still exists in the trace.
+	 */
+	if (tc->destroy_listeners) {
+		size_t i;
```
We use `uint64_t` for indexes now.

eepp commented on lib/trace-ir/trace-class.c:
```diff
@@ -63,6 +68,26 @@ void destroy_trace_class(struct bt_object *obj)
+						struct bt_trace_class_destroy_listener_elem, i);
+
+			if (elem.func) {
+				elem.func((void *) tc, elem.data);
```
Do you really need this `void *` cast?

eepp commented on include/babeltrace/trace-ir/trace-class-internal.h:
```diff
@@ -63,6 +63,8 @@ struct bt_trace_class {
 
 	bool assigns_automatic_stream_class_id;
+	GArray *destroy_listeners;
+	bool in_remove_listener;
```
I believe you can remove this member and everything using it since we don't have remove listeners anymore?

eepp commented on lib/trace-ir/trace.c:
```diff
@@ -72,24 +71,24 @@ void destroy_trace(struct bt_object *obj)
+						struct bt_trace_destroy_listener_elem, i);
+
+			if (elem.func) {
+				elem.func((void *) trace, elem.data);
```
Is this `void *` cast necessary?

eepp commented on lib/trace-ir/trace.c:
```diff
@@ -72,24 +71,24 @@ void destroy_trace(struct bt_object *obj)
-
-			if (elem.removed) {
-				elem.removed((void *) trace, elem.data);
+		BT_LIB_LOGV("Calling trace destroy listener(s): %!+t", trace);
```
Use `uint64_t` above.

eepp commented on lib/trace-ir/trace.c:
```diff
@@ -231,78 +230,46 @@ const struct bt_stream *bt_trace_borrow_stream_by_id_const(
 	}
 
-	BT_LIB_LOGV("Added \"trace is static\" listener: "
+	BT_LIB_LOGV("Added \"trace is destroy\" listener: "
```
_trace destroy_

Now that I think about it, _destroy_ is not a noun. Do you think we should use _trace destruction_ and _trace class destruction_ everywhere, including the type definitions and function names, instead? We can't use the past participle (_trace class destroyed_) because when the listener is called, the object is not destroyed yet (it's in the process of being destroyed).

frdeso replied:
I like destruction listener.

frdeso replied:
Should we change that term in component-class* files as well?

eepp replied:
> Should we change that term in component-class* files as well?

That's internal so I don't really care, it's up to you.

eepp commented on plugins/lttng-utils/bin-info.c:
```diff
@@ -143,14 +368,26 @@ int bin_info_set_build_id(struct bin_info *bin, uint8_t *build_id,
+	 * that what was recorded in the trace.
+	 */
+	bin->file_build_id_matches = is_build_id_matching(bin);
+	if(!bin->file_build_id_matches) {
```
Space.

eepp commented on plugins/lttng-utils/trace-ir-mapping.h:
```diff
@@ -0,0 +1,230 @@
+#include "debug-info.h"
+
+/* Used to resolve field paths for dynamic arrays and variant field classes. */
+struct field_class_resolving_context {
```
For each pointer member for all the structures below, please specify (comment) if the pointed object is weak or owned by the structure.

For hash tables, specify this for both the key and the value.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1072 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	upstream_iterator_ret_status =
+		bt_self_component_port_input_message_iterator_next(
+				upstream_iterator, &local_msgs, &local_count);
+	if (upstream_iterator_ret_status < 0) {
```
The only valid status with which you can read the returned messages is `BT_MESSAGE_ITERATOR_STATUS_OK`.

Then if it's something else, you can go to `error` and convert the status as you do, but I would rename the `error` label for this case because some statuses are not errors per se (`BT_MESSAGE_ITERATOR_STATUS_AGAIN`, `BT_MESSAGE_ITERATOR_STATUS_END`).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1072 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+
+	for (i = 0; i < local_count; i++) {
+		out_message = handle_message(debug_info_msg_iter, local_msgs[i]);
+		if (out_message) {
```
In which case would `out_message` be `NULL`? Would this mean an error?

eepp reviewed (changes requested):
See new comments.

Also, could you implement the message iterator's "seek beginning" method? Then we can place an `flt.utils.trimmer` before or after an `flt.lttng-utils.debug-info` component. You just need to reset the whole iterator's state in this method, and call `bt_self_component_port_input_message_iterator_can_seek_beginning()` on the upstream message iterator. You must also implement the "can seek beginning" method which just follows the result of `bt_self_component_port_input_message_iterator_can_seek_ns_from_origin()`.

eepp commented on plugins/lttng-utils/trace-ir-mapping.c:
```diff
@@ -0,0 +1,681 @@
+	BT_ASSERT(md_maps);
+	BT_ASSERT(in_stream_class);
+	BT_ASSERT(!borrow_mapped_stream_class(md_maps, in_stream_class));
+	BT_ASSERT(bt_stream_class_default_clock_is_always_known(in_stream_class));
```
I didn't mean to assert this. If this returns `BT_FALSE`, then the whole iterator must fail, as it could be possible to have unknown clock snapshots in the future.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+						out_event_class, out_packet,
+						bt_clock_snapshot_get_value(cs));
+		} else {
+			BT_LOGE_STR("Unsupported unknown clock snapshot.");
```
Since you validate that the stream clocks are always known, this should never happen. You can `BT_ASSERT(cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN)` here.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+			goto error;
+		}
+	} else {
+		out_message = bt_message_event_create(debug_it->input_iterator,
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	BT_ASSERT(out_stream);
+
+	/* Create an output stream beginning message. */
+	out_message = bt_message_stream_beginning_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	BT_ASSERT(out_stream);
+
+	/* Create an stream end output message. */
+	out_message = bt_message_stream_end_create(debug_it->input_iterator,
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+					in_message, &cs);
+
+		/* Create an packet beginning output message. */
+		if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
Assert this condition.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+			goto error;
+		}
+	} else {
+		out_message = bt_message_packet_beginning_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+					in_message, &cs);
+
+		/* Create an packet end output message. */
+		if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
Assert this.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+			goto error;
+		}
+	} else {
+		out_message = bt_message_packet_end_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		/* Create an packet end output message. */
+		if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
+			out_message =
+				bt_message_packet_end_create_with_default_clock_snapshot(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		/* Create an packet beginning output message. */
+		if (cs_state == BT_CLOCK_SNAPSHOT_STATE_KNOWN) {
+			out_message =
+				bt_message_packet_beginning_create_with_default_clock_snapshot(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	out_stream = trace_ir_mapping_borrow_mapped_stream(debug_it->ir_maps, in_stream);
+	BT_ASSERT(out_stream);
+
+	out_message = bt_message_stream_activity_beginning_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	out_stream = trace_ir_mapping_borrow_mapped_stream(debug_it->ir_maps, in_stream);
+	BT_ASSERT(out_stream);
+
+	out_message = bt_message_stream_activity_end_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		begin_cs_value = bt_clock_snapshot_get_value(begin_cs);
+		end_cs_value = bt_clock_snapshot_get_value(end_cs);
+
+		out_message = bt_message_discarded_events_create_with_default_clock_snapshots(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+					debug_it->input_iterator, out_stream,
+					begin_cs_value, end_cs_value);
+	} else {
+		out_message = bt_message_discarded_events_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+					debug_it->input_iterator, out_stream,
+					begin_cs_value, end_cs_value);
+	} else {
+		out_message = bt_message_discarded_packets_create(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		begin_cs_value = bt_clock_snapshot_get_value(begin_cs);
+		end_cs_value = bt_clock_snapshot_get_value(end_cs);
+
+		out_message = bt_message_discarded_packets_create_with_default_clock_snapshots(
```
This can fail (no memory).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	BT_ASSERT(local_count <= capacity);
+
+	for (i = 0; i < local_count; i++) {
+		out_message = handle_message(debug_info_msg_iter, local_msgs[i]);
```
`handle_message()` can fail: see all the _can fail_ comments in `handle_*()` functions.

If it fails, don't forget to put the messages that were already created and stored into `local_msgs`.

When this function fails because of no memory, return `BT_SELF_MESSAGE_ITERATOR_STATUS_NOMEM`.

frdeso commented:
I applied the changes.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1157 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+			bt_message_stream_activity_beginning_borrow_default_clock_snapshot_const(
+					in_message, &cs);
+
+		if (cs_state == BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_KNOWN) {
```
Sorry, I didn't see this last time.

This block is okay if `cs_state` is `BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_KNOWN`, but you also need to handle `BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_INFINITE`, also setting the output message's clock snapshot state to this using `bt_message_stream_activity_beginning_set_default_clock_snapshot_state()` (the default is `BT_MESSAGE_STREAM_ACTIVITY_CLOCK_SNAPSHOT_STATE_UNKNOWN`).

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1157 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	default_cc = bt_stream_class_borrow_default_clock_class_const(
+			bt_stream_borrow_class_const(in_stream));
+
+	if (default_cc) {
```
See my other comment in `handle_stream_act_begin_message()`.

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -726,22 +884,1157 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+		bt_message_put_ref(input_msgs[curr_msg_idx]);
+	}
+
+end:
```
Suggestion: I usually put the error label before `end`, with a `goto end` here (success), and the error block continuing to the end block. So that the return statement is at the end.

eepp commented on plugins/lttng-utils/trace-ir-mapping.c:
```diff
@@ -0,0 +1,684 @@
+			== BT_FALSE) {
+		BT_LOGE("Stream class default clock class is not always "
+			"known: in-sc-addr=%p", in_stream_class);
+		goto end;
```
Perfect!

frdeso commented:
changed applied.
